From bc7658182f1922d49f33acf614f408a9d3f1f9f2 Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Mon, 25 Feb 2013 07:21:54 +0000
Subject: Call fdatasync() before renaming files into place

In order to implement fully atomic upgrades, the OSTree system expects
that "triggers" such as this tool ensure that any data they generate
are on durable storage.  Thus, we use fdatasync() before calling
rename().

https://bugs.freedesktop.org/show_bug.cgi?id=61472
---
(limited to 'update-mime-database.c')

diff --git a/update-mime-database.c b/update-mime-database.c
index 305dcbd..bcf98ac 100644
--- a/update-mime-database.c
+++ b/update-mime-database.c
@@ -16,6 +16,7 @@
 #include <libxml/tree.h>
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <fcntl.h>
 
 #define XML_NS XML_XML_NAMESPACE
 #define XMLNS_NS "http://www.w3.org/2000/xmlns/"
@@ -941,6 +942,7 @@ static gboolean atomic_update(const gchar *pathname, GError **error)
 	gboolean ret = FALSE;
 	gchar *new_name = NULL;
 	int len;
+	int fd;
 
 	len = strlen(pathname);
 
@@ -948,6 +950,25 @@ static gboolean atomic_update(const gchar *pathname, GError **error)
 
 	new_name = g_strndup(pathname, len - 4);
 
+#ifdef HAVE_FDATASYNC
+	fd = open(pathname, O_RDONLY);
+	if (fd == -1)
+	{
+		set_error_from_errno(error);
+		goto out;
+	}
+	if (fdatasync(fd) == -1)
+	{
+		set_error_from_errno(error);
+		goto out;
+	}
+	if (close(fd) == -1)
+	{
+		set_error_from_errno(error);
+		goto out;
+	}
+#endif
+
 #ifdef _WIN32
 	/* we need to remove the old file first! */
 	remove(new_name);
--
cgit v0.9.0.2-2-gbebe
